### PR TITLE
feat: add Magic Sea on IOTA

### PR DESCRIPTION
**Pull request by Jeff-CCH**
- [x] add magic sea
- [x] test
- [x] yarn changeset

### CODE REVIEW

File: src/logics/magicsea/logic.swap-token.ts [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+    for (let i = 0; i < addresses.length; i++) {
+      const pairAddress = addresses[i];
+      if (!pairAddress) {
+        results.push(['0', '0']);
+      } else {
+        const pair = MagicSeaPair__factory.connect(pairAddress, this.provider);
+        const [reserve0, reserve1] = await pair.getReserves();
+        results.push([reserve0.toString(), reserve1.toString()]);
+      }
+    }
```
Comment (LuPoYi) [inline src/logics/magicsea/logic.swap-token.ts:225]: should we consider using Multicall3 to reduce the number of contract calls. Currently, one `quote` will trigger 16 contract calls? 
8 times for `getPairAddresses`
8 times for `await pair.getReserves();`

File: src/logics/magicsea/logic.swap-token.test.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+        expect(routerLogic.wrapMode).to.eq(
+          input.token.isNative
+            ? core.WrapMode.wrapBefore
+            : output.token.isNative
+            ? core.WrapMode.unwrapAfter
+            : core.WrapMode.none
+        );
+        expect(routerLogic.approveTo).to.eq(constants.AddressZero);
+        expect(routerLogic.callback).to.eq(constants.AddressZero);
```
Comment (LuPoYi) [inline src/logics/magicsea/logic.swap-token.test.ts:151]: by the chance, could you explain what `approveTo` and `callback` are for?

Comment (Jeff-CCH) [reply src/logics/magicsea/logic.swap-token.test.ts:151]: `approveTo` is for protocols that the entry contract is not the same as the token handling contract.
`callback` is for protocolink to be called again by the callback contract in a batch.

File: src/logics/magicsea/logic.swap-token.ts (hunk 1 above)
Comment (Jeff-CCH) [reply src/logics/magicsea/logic.swap-token.ts:225]: IMO, using multi-call to reduce calls is great but the logic would become more complicated. Whether to do the optimization or not depends on the RPC capability of the IOTA chain. What do @zodahu think?

Comment (zodahu) [reply src/logics/magicsea/logic.swap-token.ts:225]: I think we can use multicall here as get pairs is commonly used in this way, and our uni-v3 has implemented similar code: https://github.com/dinngo/protocolink-logics/blob/master/src/modules/univ3/logic.swap-token.ts#L240-L283

File: src/logics/magicsea/logic.swap-token.ts [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+    let j = 0;
+    for (let i = 0; i < multicallRecords.length; i++) {
+      if (multicallRecords[i]) {
+        const resultGetPair = resultsGetPair[j];
+        if (resultGetPair.success && resultGetPair.returnData !== '0x') {
+          const [pairAddress] = iface.decodeFunctionResult('getPair', resultGetPair.returnData);
+          if (pairAddress === constants.AddressZero) pairAddresses.push(undefined);
+          else {
+            pairAddresses.push(pairAddress);
+          }
```
Comment (LuPoYi) [inline src/logics/magicsea/logic.swap-token.ts:328]: we can simplify this into one line
```suggestion
          pairAddresses.push(pairAddress === constants.AddressZero ? undefined : pairAddress);
```

**Comment by LuPoYi** [comment]
LGTM

**Pull request merged**